src/imperative/cached_op_threadsafe.cc

```diff
@@ -0,0 +1,373 @@
+                                std::to_string(dedup_out[nodeEntry]++);
+        copy_node->inputs.emplace_back(nodeEntry);
+        if (_copy_op->attr_parser != nullptr) {
+          _copy_op->attr_parser(&(copy_node->attrs));
+        }
+        fwd_graph_.outputs.emplace_back(std::move(copy_node));
+      } else {
+        dedup_out.emplace(nodeEntry, 0);
+        fwd_graph_.outputs.push_back(nodeEntry);
+      }
+    }
+
+    const auto &idx = fwd_graph_.indexed_graph();
+    CHECK_GE(idx.input_nodes().size(), 1)
+        << "CachedOp requires at least 1 input";
+
+    std::vector<uint32_t> ref_count(idx.num_node_entries(), 0);
+    for (const auto &i : idx.input_nodes())
+      ++ref_count[idx.entry_id(i, 0)];
+    for (const auto &i : idx.outputs())
+      ++ref_count[idx.entry_id(i)];
+    for (size_t i = 0; i < idx.num_nodes(); ++i) {
+      for (const auto &j : idx[i].inputs)
+        ++ref_count[idx.entry_id(j)];
+    }
+
+    fwd_graph_.attrs["forward_ref_count"] =
+        std::make_shared<dmlc::any>(std::move(ref_count));
+  }
+
+  // Set param indices
+  {
+    const auto& indexed_graph = fwd_graph_.indexed_graph();
+    if (config_.data_indices.ndim() || config_.param_indices.ndim()) {
+      CHECK_EQ(config_.data_indices.ndim() + config_.param_indices.ndim(),
+               indexed_graph.input_nodes().size());
+    } else {
+      std::vector<uint32_t> tmp;
+      tmp.reserve(indexed_graph.input_nodes().size());
+      for (size_t i = 0; i < indexed_graph.input_nodes().size(); ++i) {
+        tmp.emplace_back(i);
+      }
+      config_.data_indices.assign(tmp.begin(), tmp.end());
+    }
+  }
+}
+
+OpStatePtr CachedOpThreadSafe::DynamicForward(const Context& default_ctx,
+                                              const std::vector<NDArray*>& inputs,
+                                              const std::vector<NDArray*>& outputs) {
+  using namespace nnvm;
+  using namespace imperative;
+
+  {
+  auto state_ptr = GetCachedOpState(default_ctx);
+  auto op_state = OpStatePtr::Create<DynamicRuntime>();
+  auto &runtime = op_state.get_state<DynamicRuntime>();
+  {
+    auto &state = state_ptr.get_state<CachedOpState>();
+    // Need to lock the mutex on the state, this allows
+    // for multi context push of ops to dependency engine.
+    // SetForwardGraph runs infer passes on graphs as well
+    // as the planmemory pass.
+    std::lock_guard<std::mutex> lock(state.mutex);
+    SetForwardGraph(&state.info, false, inputs);
+    runtime.info.fwd_graph = state.info.fwd_graph;
+  }
+  nnvm::Graph &g = runtime.info.fwd_graph;
+  const auto &idx = g.indexed_graph();
+  size_t num_inputs = idx.input_nodes().size();
+  size_t max_nodes = runtime.info.fwd_graph.indexed_graph().num_nodes();
+  runtime.op_states.resize(max_nodes);
+  auto &states = runtime.op_states;
+
+  // Allocate entries
+  // This buff is thread local and used to store intermediate
+  // nodes in the graph
+  buff.resize(idx.num_node_entries());
```

Comment:
This coding style is similar as in the backward pass, and I'm not a fan of this code style because I find it not very readable due to back and forth the indexed graph, indices, buffers and arrays. I think it can be abstracted more and make it easier to follow by factoring out in functions where the indexes don't appear. In the end it's the way a big chunk of mxnet is so this is like another drop in the ocean, but anyway.

Reply:
feel free to open a new pr for refactor.

Reply:
You are changing this code now, there's should be no need to immediately refactor in the future.

Reply:
i am just following existing style. Refactoring cached op is not in the scope of this PR.